### PR TITLE
Fix for issue #1117, temporary fix for set_ttl test on Windows, this …

### PR DESCRIPTION
**Pull request by dtacalau**
…test was crashing on Windows because, on Windows calling setsockopt on a

not connected socket is undefined behavior. More details can be found in issue #1117.

Signed-off-by: Daniel Tacalau <dst4096@gmail.com>

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
This change looks good to me, two questions:
* Does this apply to more methods on `TcpStream`?
* Can you add some documentation to `TcpStream::set_ttl` (and perhaps other methods) that it's required to receive an before using the method?

**Comment by dtacalau** [comment]
> * Does this apply to more methods on `TcpStream`?

@Thomasdezeeuw you mean other TcpStream methods which are using setsockopt? It so, there's only set_nodelay which suffers from the same issue, but its test has been already fixed.

> * Can you add some documentation to `TcpStream::set_ttl` (and perhaps other methods) that it's required to receive an before using the method?

Absolutely, ATM there are two affected: set_ttl and set_nodelay, others?

**Comment by Thomasdezeeuw** [comment]
If the problem is calling `setsockopt` on an unconnected socket then I think only `set_ttl` and `set_nodelay` are a problem. So only those two methods would need docs.

But does that same hold for `getsockopt`? Then it would also apply to `ttl` and `nodelay` (the getter methods).

**Comment by dtacalau** [comment]
> But does that same hold for `getsockopt`? Then it would also apply to `ttl` and `nodelay` (the getter methods).

Could be, will check all 4 of them.

**Comment by dtacalau** [comment]
Latest changes:
- on Windows, all 4 methods: set ttl, get ttl, set nodelay, get nodelay require a connected stream to work. I've added tests for all of them, also added documentation notes.
- added set/get ttl tests for TcpListener;
- added set/get ttl tests for UdpSocket;

File: src/net/tcp_stream.rs [hunk 1]
```diff
@@ -105,6 +105,9 @@ impl TcpStream {
     /// small amount of data. When not set, data is buffered until there is a
     /// sufficient amount to send out, thereby avoiding the frequent sending of
     /// small packets.
+    ///
+    /// Note On Windows: Make sure the stream is connected before calling this method.
```
Comment (Thomasdezeeuw) [inline src/net/tcp_stream.rs]: Can you changes this to:

```rust
    /// # Notes
    /// 
    /// On Windows make sure the stream is connected before calling this method,
    /// by receiving an (writable) event. Trying to set `nodelay` on an
    /// unconnected `TcpStream` is undefined behaviour.
```

Same for the others.

File: tests/udp_socket.rs [hunk 2]
```diff
@@ -111,6 +111,26 @@ fn smoke_test_unconnected_udp_socket(socket1: UdpSocket, socket2: UdpSocket) {
...
+
+#[test]
+fn get_ttl_without_previous_set() {
+    let socket1 = UdpSocket::bind(any_local_address()).unwrap();
+
+    // expect a get TTL to work w/o any previous set_ttl
+    socket1.ttl().expect("unable to get TTL for UDP socket");
+}
+
+
```
Comment (Thomasdezeeuw) [inline tests/udp_socket.rs]: Double empty line.

Comment (dtacalau) [reply tests/udp_socket.rs]: Done, removed.

File: src/net/tcp_stream.rs (hunk 1 above)
Comment (dtacalau) [reply src/net/tcp_stream.rs]: Done, reworded.

**Comment by dtacalau** [comment]
> Can you rebase on master and fix the last couple things, then we can merge this.

Done

File: src/net/tcp_stream.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/net/tcp_stream.rs]: You missed the Notes header.

Comment (dtacalau) [reply src/net/tcp_stream.rs]: Sorry for that, I've added it.

**Pull request merged**